src/main/host/syscall/mman.c

```diff
@@ -149,13 +149,33 @@ static int _syscallhandler_openPluginFile(SysCallHandler* sys, File* file) {
     /* Flush the buffer to the plugin. */
     process_flushPtrs(sys->process);
 
-    /* Get original flags that were used to open the file,
-       but be careful not to try re-creating or truncating it. */
-    int flags = file_getFlags(file) & ~(O_CREAT|O_EXCL|O_TMPFILE|O_TRUNC);
+    /* There are three categories of file flags: access mode, creation, and
+     * status flags. All of these flags apply to the underlying file object,
+     * except O_CLOEXEC which applies to the descriptor. */
+
+    /* From man 2 open (missing O_TTY_INIT since it's not defined) */
+    const int creationFlags =
+        O_CLOEXEC | O_CREAT | O_DIRECTORY | O_EXCL | O_NOCTTY | O_NOFOLLOW | O_TRUNC;
+
+    /* Get original flags that were used to open the file. */
+    int flags = file_getOriginalFlags(file);
+    /* Use only the file creation flags, except O_CLOEXEC. */
+    flags &= (creationFlags & ~O_CLOEXEC);
+    /* Add any file access mode and file status flags. */
+    flags |= fcntl(file_getOSBackedFD(file), F_GETFL);
+    /* Be careful not to try re-creating or truncating it. */
+    flags &= ~(O_CREAT|O_EXCL|O_TMPFILE|O_TRUNC);
+
+    /* Add the O_CLOEXEC creation flag if it's set for the descriptor. */
```

Comment:
I think this currently doesn't matter, since we have the managed thread close the file again after the mmap is complete. Maybe just as well to have to have it though in case e.g. we later decide to keep the file open.

Could probably use a comment explaining that rationale (or yours, if different) for doing it, though.

Reply:
My goal was just to open the file with the same flags as what the plugin had set, since before it only used the flags that the file had when it was opened and not the current flags. I changed this and added a comment.